File: frontend/src/types/subscription.ts
export enum SubscriptionFrequency {
  DAILY = 'daily',
  ALTERNATE_DAYS = 'alternate_days',
  WEEKLY = 'weekly',
}

export enum SubscriptionStatus {
  ACTIVE = 'active',
  PAUSED = 'paused',
  CANCELLED = 'cancelled',
}

export interface Subscription {
  id: number
  user_id: number
  product_id: number
  razorpay_subscription_id: string
  plan_frequency: SubscriptionFrequency
  start_date: string
  next_delivery_date: string
  delivery_address_id: number
  status: SubscriptionStatus
  created_at: string
  updated_at: string
  product: {
    id: number
    title: string
    sku: string
    unit_size: string
    consumer_price: number
    distributor_price: number
  }
  delivery_address: {
    id: number
    name: string
    phone: string
    address_line1: string
    address_line2: string | null
    city: string
    state: string
    postal_code: string
    country: string
  }
}

export interface CreateSubscriptionRequest {
  product_id: number
  plan_frequency: SubscriptionFrequency
  start_date: string
  delivery_address_id: number
}

export interface CreateSubscriptionResponse {
  subscription: Subscription
  razorpay_subscription_id: string
  razorpay_key_id: string
}

File: backend/app/schemas/subscription.py
"""Subscription schemas for request/response validation"""
from typing import List, Optional
from decimal import Decimal
from datetime import date, datetime
from pydantic import BaseModel, Field, field_validator
from app.models.enums import SubscriptionFrequency, SubscriptionStatus


class SubscriptionCreate(BaseModel):
    """Schema for creating a subscription"""
    product_id: int = Field(..., gt=0, description="Product ID for subscription")
    plan_frequency: SubscriptionFrequency = Field(..., description="Subscription frequency")
    start_date: date = Field(..., description="Subscription start date")
    delivery_address_id: int = Field(..., gt=0, description="Delivery address ID")
    
    @field_validator('start_date')
    @classmethod
    def validate_start_date(cls, v):
        """Validate start date is not in the past"""
        from datetime import date as date_type
        today = date_type.today()
        if v < today:
            raise ValueError('Start date cannot be in the past')
        return v


class ProductResponse(BaseModel):
    """Product response schema for subscription"""
    id: int
    title: str
    sku: str
    unit_size: str
    consumer_price: Decimal
    distributor_price: Decimal
    is_subscription_available: bool
    
    class Config:
        from_attributes = True


class AddressResponse(BaseModel):
    """Address response schema for subscription"""
    id: int
    name: str
    phone: str
    address_line1: str
    address_line2: Optional[str] = None
    city: str
    state: str
    postal_code: str
    country: str
    
    class Config:
        from_attributes = True


class SubscriptionResponse(BaseModel):
    """Subscription response schema"""
    id: int
    user_id: int
    product_id: int
    razorpay_subscription_id: str
    plan_frequency: SubscriptionFrequency
    start_date: date
    next_delivery_date: date
    delivery_address_id: int
    status: SubscriptionStatus
    product: ProductResponse
    delivery_address: AddressResponse
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class SubscriptionListResponse(BaseModel):
    """Subscription list response"""
    subscriptions: List[SubscriptionResponse]
    total: int


class RazorpaySubscriptionResponse(BaseModel):
    """Razorpay subscription creation response"""
    razorpay_subscription_id: str
    status: str
    subscription_id: int
    key_id: str


class SubscriptionPaymentVerification(BaseModel):
    """Schema for verifying subscription payment"""
    razorpay_payment_id: str
    razorpay_subscription_id: str
    razorpay_signature: str
